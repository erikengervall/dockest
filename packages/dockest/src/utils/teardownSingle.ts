import { execaWrapper } from './execaWrapper'
import { Runner } from '../@types'
import { DockestError } from '../Errors'

const stopContainerById = async ({ runner, runner: { containerId } }: { runner: Runner }) => {
  const logPrefix = '[Stop Container]'

  const command = `docker stop ${containerId}`
  await execaWrapper(command, { runner, logPrefix, logStdout: true })
}

const removeContainerById = async ({ runner, runner: { containerId } }: { runner: Runner }) => {
  const logPrefix = '[Remove Container]'

  const command = `docker rm ${containerId} --volumes`
  await execaWrapper(command, { runner, logPrefix, logStdout: true })
}

export const teardownSingle = async ({
  runner,
  runner: { containerId, dependents, serviceName },
}: {
  runner: Runner
}) => {
  if (!containerId) {
    throw new DockestError(`Invalid containerId (${containerId}) for service (${serviceName})`, { runner })
  }

  for (const dependant of dependents) {
    await teardownSingle({ runner: dependant })
  }

  await stopContainerById({ runner })
  await removeContainerById({ runner })
}
