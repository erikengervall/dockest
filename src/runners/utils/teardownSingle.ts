import execa from 'execa'

import { globalLogger, runnerLogger } from '../../loggers'

const stopContainerById = async (containerId: string, runnerKey: string): Promise<void> => {
  runnerLogger.stopContainer(runnerKey)

  try {
    const cmd = `docker stop ${containerId}`

    runnerLogger.shellCmd(cmd)
    await execa.shell(cmd)
  } catch (error) {
    globalLogger.error(`${runnerKey}: Failed to stop service container`, error)
    return
  }

  runnerLogger.stopContainerSuccess(runnerKey)
}

const removeContainerById = async (containerId: string, runnerKey: string): Promise<void> => {
  runnerLogger.removeContainer(runnerKey)

  try {
    const cmd = `docker rm ${containerId} --volumes`

    runnerLogger.shellCmd(cmd)
    await execa.shell(cmd)
  } catch (error) {
    globalLogger.error(`${runnerKey}: Failed to remove service container`, error)

    return
  }

  runnerLogger.removeContainerSuccess(runnerKey)
}

export default async (containerId: string, runnerKey: string): Promise<void> => {
  if (!containerId) {
    globalLogger.error(`${runnerKey}: Cannot teardown container without a containerId`)
    return
  }

  runnerLogger.teardown(runnerKey)
  await stopContainerById(containerId, runnerKey)
  await removeContainerById(containerId, runnerKey)
  runnerLogger.teardownSuccess(runnerKey)
}
