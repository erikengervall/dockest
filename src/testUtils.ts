import { createMockProxy } from 'jest-mock-proxy'
import execa from 'execa' // eslint-disable-line import/default
import { ZooKeeperRunner, KafkaRunner, PostgresRunner, RedisRunner, GeneralPurposeRunner } from './runners'
import { Runner } from './runners/@types'
import Logger from './Logger'
import { DEFAULT_USER_CONFIG, INTERNAL_CONFIG } from './constants'

export const mockedExecaStdout = 'getContainerId ðŸŒ®'
export const runnerCommand = 'runRunnerCommands ðŸŒ®'

/**
 * FIXME: Would love for this to work ðŸ¦Š
 */
// jest.mock('execa', () => jest.fn(() => ({ stdout: mockedExecaStdout })))
// jest.mock('./Logger')

const { values } = Object

const createDockestConfig = (opts: { runners?: Runner[] }) => {
  const { runners = [] } = opts

  for (const runner of runners) {
    runner.logger = createMockProxy()
  }

  return {
    runners,
    opts: DEFAULT_USER_CONFIG,
    jest: {},
    $: INTERNAL_CONFIG,
  }
}

export default ({ withRunnerCommands = false }) => {
  const withCmds = withRunnerCommands ? { commands: [runnerCommand] } : {}

  const zooKeeperRunner = new ZooKeeperRunner({ service: 'zookeepeer', image: 'some/image:123', ...withCmds })
  const kafkaRunner = new KafkaRunner({
    service: 'kafka',
    image: 'some/image:123',
    dependsOn: [zooKeeperRunner],
    ...withCmds,
  })
  const postgresRunner = new PostgresRunner({
    service: 'postgres',
    image: 'some/image:123',
    database: '_',
    username: '_',
    password: '_',
    ...withCmds,
  })
  const redisRunner = new RedisRunner({ service: 'redis', image: 'some/image:123', ...withCmds })
  const generalPurposeRunner = new GeneralPurposeRunner({ service: 'general', image: 'some/image:123', ...withCmds })
  const initializedRunners = {
    kafkaRunner,
    postgresRunner,
    redisRunner,
    zooKeeperRunner,
    generalPurposeRunner,
  }

  beforeEach(() => values(initializedRunners).forEach(runner => (runner.logger = createMockProxy())))

  return {
    initializedRunners,
    runners: {
      ZooKeeperRunner,
      KafkaRunner,
      PostgresRunner,
      RedisRunner,
      GeneralPurposeRunner,
    },
    execa,
    Logger,
    createDockestConfig,
  }
}
