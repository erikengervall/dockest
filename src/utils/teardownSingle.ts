import { Runner } from '../runners/@types'
import execaWrapper from './execaWrapper'

const teardownSingle = async (runner: Runner): Promise<void> => {
  const { containerId, logger } = runner

  if (!containerId) {
    logger.error('Cannot teardown container without a containerId')
    return
  }

  // Teardown runner's dependencies
  for (const depRunner of runner.runnerConfig.dependsOn) {
    const { logger } = depRunner

    logger.info('Container teardown started')
    await stopContainerById(depRunner)
    await removeContainerById(depRunner)
    logger.info('Container teardown completed', { nl: 1 })
  }

  // Teardown runner
  logger.info('Container teardown started')
  await stopContainerById(runner)
  await removeContainerById(runner)
  logger.info('Container teardown completed', { nl: 1 })
}

const stopContainerById = async (runner: Runner): Promise<void> => {
  const { containerId, logger } = runner
  logger.debug('Container being stopped')

  try {
    const command = `docker stop ${containerId}`

    await execaWrapper(command, runner)
  } catch (error) {
    logger.error('Unexpected error when stopping container')

    return
  }

  logger.debug('Container stopped')
}

const removeContainerById = async (runner: Runner): Promise<void> => {
  const { containerId, logger } = runner
  logger.debug('Container being removed')

  try {
    const command = `docker rm ${containerId} --volumes`

    await execaWrapper(command, runner)
  } catch (error) {
    logger.error('Unexpected error when removing container')

    return
  }

  logger.debug('Container removed')
}

export default teardownSingle
