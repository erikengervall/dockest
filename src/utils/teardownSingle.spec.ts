import { createMockProxy } from 'jest-mock-proxy'
import testUtils from '../testUtils'
import teardownSingle from './teardownSingle'

const { redisRunner, execa } = testUtils({})
redisRunner.containerId = 'mockContainerId'

describe('teardownSingle', () => {
  beforeEach(() => {
    // @ts-ignore
    execa.mockClear()
    redisRunner.logger = createMockProxy()
  })

  describe('happy', () => {
    it('should work', async () => {
      await teardownSingle(redisRunner)

      expect(execa).toHaveBeenCalledWith(expect.stringMatching('docker stop'), { shell: true })
      expect(execa).toHaveBeenCalledWith(expect.stringMatching('docker rm'), { shell: true })
    })
  })

  describe('sad', () => {
    it('should log and swallow teardown errors', async () => {
      const error = new Error('Unexpected teardown error')
      // @ts-ignore
      execa.mockImplementation(() => {
        throw error
      })

      await teardownSingle(redisRunner)

      expect(redisRunner.logger.info).toHaveBeenCalled()
      expect(redisRunner.logger.error).toHaveBeenCalled()
    })
  })
})
