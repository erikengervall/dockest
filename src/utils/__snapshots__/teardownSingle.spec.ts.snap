// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`teardownSingle happy should work 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "docker stop mockContainerId",
      Object {
        "shell": true,
      },
    ],
    Array [
      "docker rm mockContainerId --volumes",
      Object {
        "shell": true,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "stdout": "getContainerId ðŸŒ®",
      },
    },
    Object {
      "type": "return",
      "value": Object {
        "stdout": "getContainerId ðŸŒ®",
      },
    },
  ],
}
`;

exports[`teardownSingle sad should log and swallow teardown errors 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Container teardown started",
    ],
    Array [
      "Container teardown completed",
      Object {
        "nl": 1,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`teardownSingle sad should log and swallow teardown errors 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Unexpected error when stopping container",
      Object {
        "error": [Error: Unexpected teardown error],
      },
    ],
    Array [
      "Unexpected error when removing container",
      Object {
        "error": [Error: Unexpected teardown error],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
